fix: Update typescript definition files for core, blocks, and generators

maribethb opened the pull request:
<!--
  - Thanks for submitting code to Blockly!  Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## The basics

<!-- TODO: Verify the following, checking each box with an 'x' between the brackets: [x] -->

- [x] I branched from develop
- [x] My pull request is against develop
- [x] My code follows the [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide)
- [x] I ran `npm run format` and `npm run lint`

## The details
### Resolves

Fixes #6079 

### Proposed Changes

Updates TypeScript definition files to properly export Blockly / core and Blocks. I also propose that we cherry pick this change into master and release a patch version of v8.

#### Behavior Before Change

- TypeScript errors when importing Blockly unless you turn off TS checking for libraries (see linked issue)
- No autocomplete when importing Blockly
- No TS definitions available for generators 
- No autocomplete for generators or messages

#### Behavior After Change

- No TS errors when importing Blockly in the way we advise you to
- Autocomplete available for core library
- limited autocomplete (due to incomplete types) of generators, libraryBlocks, and messages

### Reason for Changes

Improve lives of TS developers. Right now Blockly gives TS errors when imported; this used to work so some developers are not updating until this is fixed.

### Test Coverage

Tested in several ways.

I used this test repro from Johnny: https://github.com/johnnyoshika/blockly-typescript-libcheck and removed the "skipLibCheck" from the ts-config.

Then tested that none of this gave errors:

```
import * as Blockly from 'blockly';
import Blocks from 'blockly/blocks';
import * as PHP from 'blockly/php';
import * as Dart from 'blockly/dart';
import * as Lua from 'blockly/lua';
import * as JavaScript from 'blockly/javascript';
import * as libraryBlocks from 'blockly/blocks';
import {loops} from 'blockly/blocks';

JavaScript.addReservedWords('hello');

Blockly.JavaScript.COMMENT_WRAP;

PHP.FUNCTION_NAME_PLACEHOLDER_REGEXP_;
PHP.addReservedWords('testing');
Dart.addReservedWords('hi');
Blockly.Workspace.getAll();
Blockly.JavaScript.addReservedWords('test');

Blockly.Blocks;

PHP.workspaceToCode(Blockly.getMainWorkspace());

Blockly.libraryBlocks.loops.loopTypes;
libraryBlocks.loops.loopTypes;
loops.loopTypes;

console.log(Blockly, Blockly.WorkspaceSvg, Blockly.Msg, Blockly.Blocks, Blocks, PHP, Blockly.JavaScript);
```

Also tested that

`import * as Blockly from 'blockly/core';` works as expected (gives you core blockly, but not JS)

and if you have the `allowSyntheticDefaultModules` flag enabled in your tsconfig file, you can also do `import Blockly from 'blockly';`. The reason you have to have this flag is that we do not provide a default export for the Blockly module. We export quite a lot of different things from the Blockly module and you can only have one default export. We could provide a named export if we did something like `export * as Blockly from 'core/blockly'` but the problem is that consumers would have to then use `import { Blockly } from 'blockly'` which is not what we tell them to do. All of our documentation tells them to use `import * as Blockly from 'blockly';` and this approach is compatible with that guidance. It's just a free bonus that this flag exists and allows consumers to use `import Blockly from 'blockly'`.

I also tested this with the scroll-options plugin in blockly-samples which does not use TypeScript compiler directly. After this change, VS Code properly uses intellisense, and the plugin also compiles and runs correctly still. Checking both of these options is how I am satisfied that the recommended ways of importing modules work with both the actual module (the JavaScript code) and the TypeScript definitions (because otherwise it's possible that the TS compiler is happy yet the actual JavaScript does not exist)


### Why is the JavaScript generator different from the others?

You might have noticed above that I referenced the JS generator with `Blockly.JavaScript` and the PHP generator with a different import statement and reference `PHP` directly. This is because if you `import * as Blockly from 'blockly';` then the JavaScript generator is included in the `*` (import everything from the given module), so the JS generator is available on the name you gave it, Blockly. 

The other generators are also included in the default import in node, but not in the browser :/ however the fact that they are included in the default node package is not documented so I have not added it to the type in this PR. To reference the other language generators you can use `import * as PHP from 'blockly/php';` which will get you proper type checking.

You also have the option of just doing `import 'blockly/php'` which as a side effect will make `Blockly.PHP` available as a module, but TypeScript won't know about it. This is what we currently recommend to people in the README, and we just don't mention TS at all. I believe most people currently use `(Blockly.PHP as any).workspaceToCode()` to avoid this issue. With these updated definitions you can now use the above syntax and get proper type checking.

### What's up with the `export =` syntax?

I originally tried other solutions where I don't use the `export = ` syntax and I use either a named or default export of a constant named e.g. `PHP` so something like `export const PHP: Blockly.Generator` but this adds a layer of indirection to the resulting module imports. It's complicated because for the actual JavaScript code to work you *must* import the module as `import * as PHP from 'blockly/php` (or just `import 'blockly/php'`) but then that fails the TypeScript compiler. The TypeScript compiler is only happy with this definition syntax if you import it like `import { PHP } from 'blockly/php'` but then the actual module loader is unhappy because that's not how the JavaScript module works. So you end up going in circles here. Therefore I just left the `export =` syntax alone.

### What's going on with Blockly.Msg and the locales?

Similar to the above, I tried various strategies but couldn't arrive at a solution that satisfied both the module loader and the TypeScript compiler. So the current status is you have no types for Blockly.Msg still or any of the locales.

### Documentation

We could update the README to instruct TypeScript users to use `import * as PHP from 'blockly/php';` instead of `import 'blockly/php';` but I haven't done that in this PR yet. I wanted to be sure of this solution before updating our documentation.

### Additional Information

<!-- Anything else we should know? -->

maribethb committed:
fix: update definitions for core and blocks

maribethb committed:
fix: update javascript definition

maribethb committed:
fix: update JS definition to work with both import types

cpcallen commented on typings/blocks.d.ts:
```diff
@@ -11,5 +11,4 @@
 
-import * as Blockly from './core';
-export = Blockly.Blocks;
+export { Blocks } from './core';
```
Hm…  I think this is wrong.  If I understand correctly, this file is supposed to be typings for the stuff in `blocks/`—i.e., the `Blockly.libraryBlocks` modules—not the single dictionary object exported by `core/blocks.js`.

Notably, while I think it was believed that `import * as Blocks from 'blockly/blocks'` would get you a copy of the `Blockly.Blocks` dictionary object, that was AFAIK not in fact the case (due to bugs) at the time we transitioned to `goog.module`, if ever.

In any case, the blocks modules got their own more useful set of exports via PRs #5900 and #5908, and were renamed to `Blockly.libraryBlocks` in #5953.  Some appropriate typings should probably have been added along the way but regrettably [the author](https://github.com/cpcallen) of those PRs neither thought to do so nor knew how to.

maribethb replied:
OK, I was definitely under the impression that importing `blockly/blocks` got you the `Blockly.Blocks` library object but loaded with the library blocks into it already. If that's not the case. I don't think we currently have type definitions for anything exported under the `libraryBlocks` modules, so I don't know what to put here (as again it was outside the scope of this to add new definitions from scratch, I'm just trying to resolve the errors). I think we'd get actual type definitions when we move fully to TS.

So for now I can define `libraryBlocks` as ... something, which will allow people to use `Blockly.libraryBlocks` but won't provide deeper type checks? `Blockly.Blocks` will continue to work as that type is defined in core.

cpcallen commented on typings/index.d.ts:
```diff
@@ -14,9 +14,8 @@
+import { Generator } from 'core/blockly';
+export * from './core';
+export * from './blocks';
+export const JavaScript: Generator;
```
1. Should we provide typings for the other language generators too?  They do all get loaded if you import `'blockly/blocks'` in node.js, even if they don't in the browser (ugh).
0. I believe the actual typings for generator module default exports are considerably more elaborate than just a simple `Generator` instance—for example, `generators/javascript.js` adds a bunch of `.ORDER_*` properties that are not on base `Generator` instances.  Do we care about such things?  ("It would be nice but is not important at the moment" would be a perfectly reasonable reply.)

maribethb replied:
1. The generators are loaded if you import `blockly/blocks`? why / where is this done? I don't see anything about it in the various `blocks`-related files... why would we do this and do we intend to keep doing it? in any case then I guess we should, but they'd have to go into the blocks definition file, which gets complicated as then you would have to reference the generators off of the import object from the blocks which makes very little sense imo.
2. Yeah, I know there's extra stuff for them but we don't have existing type definitions for the extra stuff and I don't have time to add them now. I think this will be resolved when we just move everything to TS.

maribethb replied:
Okay I see in `node.js` where we import all of the generators, so I'm going to assume you had a typo and meant "if you import `blockly' in node.js". In that case then yes I can add type definitions for the other generators to the index.d.ts, but also the fact that we do that is contrary to what our documentation says... do you think we plan to keep doing that? if we don't then i'd rather not add the definitions here because i don't want to encourage people to rely on these secretly imported modules we haven't previously told anyone about.

cpcallen reviewed (changes requested):
I think that `blocks.d.ts` is almost certainly wrong and should probably be fixed.  I'm slightly dubious about some of the rest but we can probably live with it.

cpcallen commented:
> We could update the README to instruct TypeScript users to use `import * as PHP from 'blockly/php';` instead of `import 'blockly/php';` but I haven't done that in this PR yet. I wanted to be sure of this solution before updating our documentation

That sounds like a good idea to me—only I think we should hold off until after TS migration forces a change to (probably) `import {PHP} from 'blockly/php` (or, if we use `export default` in `php.js`, then `import PHP from blockly/php`).

johnnyoshika commented:
@maribethb Thanks for using my test repro to test `skipLibCheck` 😀

maribethb committed:
fix: update typings for blocks

maribethb commented:
OK some updates:
1. Changed `blocks` to export some constants corresponding to the modules of `libraryBlocks` but typed them as any. This isn't ideal obviously but we'll have much better types once everything is actually in TypeScript, this gets it unblocked for now. Updated the test cases in the PR description.
2. Intellisense actually does work on the generators and stuff. In VS Code I opened the source code "go to definition" for the modules where intellisense wasn't working, and that caused it to recompute or whatever and suddenly I have working autocomplete. So that's nice.
3. After thinking about it more I am definitely in favor of not adding the other generators to the default package as that gives us maximum flexibility as we rework the modules in the upcoming release, since we have never documented their inclusion previously.

Also @johnnyoshika your repo was super helpful. When we get these fixes merged in would you be willing to test a beta release in your actual repository to ensure the fixes work before we push out a patch?

johnnyoshika commented:
@maribethb yes I'd be happy to help

maribethb commented:
Hey @johnnyoshika I just published blockly `8.0.3-beta.0` with the updated type definitions, would love to have your feedback! Mostly looking for whether the basic errors you were previously getting (such as "This module can only be referenced with ECMAScript imports/exports by turning on the 'allowSyntheticDefaultImports' flag and referencing its default export.") are resolved, the types still aren't perfect but any feedback would be great!

johnnyoshika commented:
@maribethb Thanks for addressing this problem. I can confirm that the `skipLibCheck` and `allowSyntheticDefaultImports` problems are now fixed with this beta release. However, other "type" problems now appear with the `Blockly.JavaScript` generator. I think this is because `Blockly.JavaScript` is now properly typed, whereas it was simply `any` in blockly@8.0.2. This is a step in the right direction, but because the types aren't fully defined, it'll result in build errors like these:

![image](https://user-images.githubusercontent.com/504505/171559110-8a86458d-529e-4635-8adc-d0ca4108a368.png)

You can reproduce these errors using a couple of new branches I created in my repo: https://github.com/johnnyoshika/blockly-typescript-libcheck

```
git checkout blockly-javascript
npm ci
npm run build
```

`blockly-javascript` branch uses blockly@8.0.2 and everything compiles properly. Now try this:

```
git checkout blockly-8.0.3-beta.0
npm ci
npm run build
```

`blockly-8.0.3-beta.0` branch uses the beta release. Trying to compile this branch generates the errors shown in my screenshot above.

We can temporarily get around this problem by using `// @ts-ignore` every time we use `Blockly.JavaScript` in a TypeScript project to instruct the compiler to suppress these errors. There may also be a better way to work around this problem that I'm not aware of, but I think this is a workable solution for now until all of the types get sorted in Blockly.

maribethb commented:
8.0.3-beta.1 has been published now!

johnnyoshika commented:
@maribethb it seems there's another type regression. I don't know when it was introduced, but it's now related to `Blockly.inject`.

To reproduce, you can use my repo again: https://github.com/johnnyoshika/blockly-typescript-libcheck

First test with blockly@8.0.1:

```
git checkout [blockly-inject
npm ci
npm run build
```

Everything will build fine.

Now try with blockly-8.0.3-beta.1:

```
git checkout blockly-8.0.3-beta.1
npm ci
npm run build
```

It'll result in this error:

![image](https://user-images.githubusercontent.com/504505/172254529-fa261cd6-744b-49d5-a14a-b9f0c8dcb470.png)

The type definition is expecting `() => void` as the second parameter to `Blockly.inject`:

![image](https://user-images.githubusercontent.com/504505/172254874-c76cfad3-cc6c-4c6e-9078-1590bc45a22d.png)

...whereas it should allow an object with these options: https://developers.google.com/blockly/guides/configure/web/configuration_struct

johnnyoshika commented:
@maribethb How can I create a local build of `blockly` so that I can use `npm link` to test changes locally without you having to publish to npm? I can run blockly with `npm start` and I tried `npm run build`, but that doesn't seem to produce the same output as the published files. I watched your video here https://youtu.be/vuW-L6ettx0?t=532, but I couldn't find a section on building locally. Thanks!

maribethb commented:
You can do `npm run package` which will package up everything the same as it will be on npm into the `dist` directory, then `cd dist` and `npm link`. Important that you link from the `dist` directory to simulate what you'd get from npm. Slightly more info here: https://developers.google.com/blockly/guides/plugins/debugging#method_1_npm_link (this is focused on plugins but same principles apply to link to any other project outside blockly). If you are only changing files in the typings/ directory you can skip the build step, but you will need to rerun `npm run package` whenever you make a change.

Regarding the type regression, that was almost certainly introduced by v8 but hidden by the errors preventing type checking at all. I think it's related to how Blockly.Options was declared somewhat weirdly as a special case before and we didn't update how that was used. I do want to fix up our definitions, but I also have to draw the line somewhere for how much we should fix in this patch release in order to unblock people. So I think best move is to release what we've currently got as a patch, and devs will have to cast to any when injecting :/ it's not ideal but at least it's contained (hopefully) to a small area, and it's better than the current status of throwing an error you can't cast to get around. we will continue to accept contributions that improve our types as well!

johnnyoshika commented:
@maribethb I agree. I think your changes are ready for release.

As for contributing type updates, would the Blockly team be interested in small PRs as I come across type errors (e.g. `Blockly.inject`)? Or is a larger holistic type update preferred?

Thanks for explaining the `npm link` strategy. This will be very helpful.

BTW, I contributed a type update to the `blockly-samples` repo but it's been sitting in PR for 2 weeks. Hint hint 😉 https://github.com/google/blockly-samples/pull/1143

maribethb commented:
The patch release v8.0.3 is now released.

@johnnyoshika Either way would be fine! It's probably slightly easier to release patch releases in bigger chunks, but small changes are fine in develop. Before I send you down a rabbit hole though, know that in our next release we plan to convert the whole codebase to TypeScript rather than relying on these clumsily generated declaration files. So don't spend too much time making fixes directly in the `.d.ts` files as there is currently a lot of work being done behind the scenes to convert to TS. After that is done I anticipate there will still be a need for type refinement as I believe we are currently "resolving" a lot of compiler errors by casting to `any` temporarily to unblock the migration, so I expect improving that will be the gameplan moving forward.
